### PR TITLE
store: Support table scan, index scan and selection with DAG.

### PR DESCRIPTION
**Pull request by zimulala**
`SendCopReqNew` is used for testing now.

### CODE REVIEW

File: executor/executor_test.go [hunk 1]
```diff
@@ -249,6 +249,36 @@ func (s *testSuite) TestSelectLimit(c *C) {
...
+	}()
+	tk.MustExec("use test")
+	tk.MustExec("create table select_dag(id int not null default 1, name varchar(255));")
+
+	tk.MustExec("insert INTO select_dag VALUES (1, \"hello\");")
+	tk.MustExec("insert INTO select_dag VALUES (2, \"hello\");")
+	tk.MustExec("insert INTO select_dag VALUES (3, \"hello\");")
+	tk.CheckExecResult(1, 0)
+
+	tk.MustExec("begin")
```
Comment (hanfei1991) [inline executor/executor_test.go]: No need to add `begin` and `commit`

File: store/tikv/mock-tikv/cop_handler_dag.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+			return nil, errors.Trace(err)
+		}
+		currExec = &selectionExec{
+			Selection: curr.Selection,
+			eval:      ctx.eval,
+			columns:   cols,
+			sc:        ctx.sc,
+		}
+	default:
+		// TODO: support other types.
```
Comment (hanfei1991) [inline store/tikv/mock-tikv/cop_handler_dag.go]: We should raise an error when meeting the unexpected type.

File: store/tikv/mock-tikv/cop_handler_dag.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	"github.com/pingcap/tidb/distsql"
+	"github.com/pingcap/tidb/distsql/xeval"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+// MockDAGRequest is used for testing now.
+var MockDAGRequest *bool
```
Comment (hanfei1991) [inline store/tikv/mock-tikv/cop_handler_dag.go]: Why a pointer?

File: store/tikv/mock-tikv/executor.go [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+		return 0, nil, errors.Trace(err)
+	}
+	row, err := HandleRowData(e.Columns, e.colTps, handle, val)
+	if err != nil {
+		return 0, nil, errors.Trace(err)
+	}
+	return handle, row, nil
+}
+
+func (e *tableScanExec) getRowFromRange(ran kv.KeyRange) (int64, map[int64][]byte, error) {
```
Comment (hanfei1991) [inline store/tikv/mock-tikv/executor.go]: Can you reuse the code of seeking key in `IndexScanExec` ?

File: store/tikv/mock-tikv/executor.go [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+			return 0, nil, errors.Trace(err)
+		}
+		if boolResult == 1 {
+			return handle, row, nil
+		}
+	}
+}
+
+// HandleRowData deals with raw row data:
+//	1. Decodes row from raw byte slice.
```
Comment (hanfei1991) [inline store/tikv/mock-tikv/executor.go]: Only have item 1?

File: store/tikv/mock-tikv/rpc.go [hunk 6]
```diff
@@ -399,14 +401,85 @@ func (c *RPCClient) SendCopReq(ctx goctx.Context, addr string, req *coprocessor.
 	default:
 	}
 
+	if *MockDAGRequest {
+		if req.GetTp() == kv.ReqTypeSelect || req.GetTp() == kv.ReqTypeIndex {
+			resp, isContinue, err := c.SendCopReqNew(addr, req, timeout)
```
Comment (hanfei1991) [inline store/tikv/mock-tikv/rpc.go]: When `isContinue` is true?

File: executor/executor_test.go (hunk 1 above)
Comment (zimulala) [inline executor/executor_test.go]: It's added as the old test code.

File: store/tikv/mock-tikv/cop_handler_dag.go (hunk 3 above)
Comment (zimulala) [inline store/tikv/mock-tikv/cop_handler_dag.go]: It needs to change by `tikv.MockDAGRequest`.

File: store/tikv/mock-tikv/executor.go (hunk 4 above)
Comment (zimulala) [inline store/tikv/mock-tikv/executor.go]: No.

File: store/tikv/mock-tikv/executor.go (hunk 5 above)
Comment (zimulala) [inline store/tikv/mock-tikv/executor.go]: Yep.

File: executor/executor_test.go (hunk 1 above)
Comment (hanfei1991) [inline executor/executor_test.go]: Then you can remove it , it's useless.

File: store/tikv/mock-tikv/executor.go (hunk 4 above)
Comment (hanfei1991) [inline store/tikv/mock-tikv/executor.go]: The seeking logic is the same. I think you can do it.

File: store/tikv/mock-tikv/rpc.go (hunk 6 above)
Comment (zimulala) [inline store/tikv/mock-tikv/rpc.go]: This is written according to the old code, and it's used for testing.

File: store/tikv/mock-tikv/executor.go (hunk 4 above)
Comment (zimulala) [inline store/tikv/mock-tikv/executor.go]: The `e.Seek` isn't the same.

File: store/tikv/mock-tikv/executor.go [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+	seekKey   []byte
+
+	src executor
+}
+
+func (e *tableScanExec) SetSrcExec(exec executor) {
+	e.src = exec
+}
+
+func (e *tableScanExec) Close() error {
```
Comment (hanfei1991) [inline store/tikv/mock-tikv/executor.go]: Close isn't used.

File: store/tikv/mock-tikv/rpc.go (hunk 6 above)
Comment (hanfei1991) [inline store/tikv/mock-tikv/rpc.go]: remove it ~

File: store/tikv/mock-tikv/cop_handler_dag.go (hunk 2 above)
Comment (zimulala) [inline store/tikv/mock-tikv/cop_handler_dag.go]: It only three types now.

**Comment by zimulala** [comment]
PTAL @coocood @shenli

File: store/tikv/mock-tikv/cop_handler.go [hunk 8]
```diff
@@ -476,13 +459,13 @@ func (h *rpcHandler) getRowData(value []byte, colTps map[int64]*types.FieldType)
 }
 
 // Put column values into ctx, the values will be used for expr evaluation.
-func (h *rpcHandler) setColumnValueToCtx(ctx *selectContext, handle int64, row map[int64][]byte, cols map[int64]*tipb.ColumnInfo) error {
+func setColumnValueToCtx(eval *xeval.Evaluator, handle int64, row map[int64][]byte, cols map[int64]*tipb.ColumnInfo) error {
```
Comment (coocood) [inline store/tikv/mock-tikv/cop_handler.go]: Why change this method?

File: store/tikv/mock-tikv/cop_handler_dag.go (hunk 3 above)
Comment (coocood) [inline store/tikv/mock-tikv/cop_handler_dag.go]: We have the flag in `store/tikv` already.
We can handle the request based on the request type.

File: store/tikv/mock-tikv/executor.go [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+			return nil, errors.New("Miss column")
+		}
+		values[col.GetColumnId()] = []byte{codec.NilFlag}
+	}
+
+	return values, nil
+}
+
+// SetColumnValueToEval puts column values into evaluator, the values will be used for expr evaluation.
+func SetColumnValueToEval(eval *xeval.Evaluator, handle int64, row map[int64][]byte, cols map[int64]*tipb.ColumnInfo) error {
```
Comment (coocood) [inline store/tikv/mock-tikv/executor.go]: make it unexported.
And it seems like it's the same as `setColumnValueToCtx`.

File: store/tikv/mock-tikv/cop_handler.go (hunk 8 above)
Comment (zimulala) [inline store/tikv/mock-tikv/cop_handler.go]: I need to use this function, but their arguments are different.

Comment (coocood) [inline store/tikv/mock-tikv/cop_handler.go]: This is the same as `setColumnValueToEval`

**Comment by coocood** [comment]
LGTM

**Pull request merged**